### PR TITLE
Improve speed of tesseract by optimising for intSimdMatrix case

### PR DESCRIPTION
**Pull request by robinwatts**
The single largest consumer of time for tesseract is the MatrixDotVector routine. Lots of work has been done on this already, including optimised integer based versions with SIMD implementations.

The last stage of these implementations involve biasing, scaling and writing the results out to memory.

Traditionally that stage has always had to be done outside of the SIMD. By tweaking the calculation slightly (so as to give the same results), we can push that calculation into the SIMD section. The downside to doing that is that we buffer overrun the biases and scales as we read them and we buffer overrun the outputs as we write them.

It turns out that the storage for the biases is already rounded, so the buffer overrun there is no problem. By adjusting the storage for the scales and the outputs, we can similarly ensure that the buffer overruns there are no problem either.

This results in a noticeable speedup. Running tesseract on 8087_054.3G.tif, runtime drops from 2.681s (real), 8.582 (user) to 2.582 (real), 8.174 (user) for identical results (times from an x86 box).

### CODE REVIEW

**Comment by robinwatts** [comment]
I don't believe the test failures are my fault. Looks like a CMake issue?

**Comment by stweil** [comment]
The cmake version which is used for CI is too old. @egorpugin, do you have a suggestion how to fix that?

**Comment by robinwatts** [comment]
Timings on my Raspberry pi. OLD: 23.445(real) 48.279(user) -> NEW: 23.016(real) 46.990(user)

If we could work in floats rather than doubles, then more could be done within the SIMD on NEON.

File: src/arch/intsimdmatrix.h
Comment (stweil) [inline src/arch/intsimdmatrix.h:65]: This breaks compilation of `make check`.

File: include/tesseract/genericvector.h
Comment (stweil) [inline include/tesseract/genericvector.h:69]: I would like to replace class `GenericVector` by standard classes, but each new function makes that more difficult. Try to avoid changing `genericvector.h`.

File: src/arch/intsimdmatrix.cpp
Comment (stweil) [inline src/arch/intsimdmatrix.cpp:92]:
```suggestion
    v[i] = (total + wi[num_in] * INT8_MAX) * scales[i];
```

File: src/arch/intsimdmatrixneon.cpp
Comment (stweil) [inline src/arch/intsimdmatrixneon.cpp:121]: The type casts here and below are not needed.

File: src/lstm/weightmatrix.cpp
Comment (stweil) [inline src/lstm/weightmatrix.cpp:185]:
```suggestion
    scales_.scale(1.0 / INT8_MAX);
```

File: src/lstm/weightmatrix.cpp
Comment (stweil) [inline src/lstm/weightmatrix.cpp:205]:
```suggestion
    scales_.scale(1.0 / INT8_MAX);
```

File: include/tesseract/genericvector.h
Comment (robinwatts) [reply include/tesseract/genericvector.h:69]: How can I avoid changing GenericVector if I need GenericVector to do something different?

I need to be able to ensure that while a GenericVector of size n still has size n, it has enough space allocated so that I can access elements up to "n rounded up a bit" without causing SEGVs.

If that can be done with the existing GenericVector interface, then please tell me how, and I'll gladly change it.

Comment (stweil) [reply include/tesseract/genericvector.h:69]: Wouldn't it be sufficient to use the existing API, calling `reserve()` and `truncate()`, for example?

Comment (robinwatts) [reply include/tesseract/genericvector.h:69]: It would. I'll move over to that. Nice.

**Comment by robinwatts** [comment]
New version, with your suggestions applied, and rebased on master. Apologies for the delay!

**Comment by zdenop** [comment]
@stweil: could it be merged now or do you require more changes?

**Comment by stweil** [comment]
> could it be merged now or do you require more changes?

It's like to run some more tests before merging.

File: src/arch/intsimdmatrix.h
Comment (stweil) [reply src/arch/intsimdmatrix.h:65]: @robinwatts, `make check` still does not compile.

File: src/arch/intsimdmatrixavx2.cpp
Comment (stweil) [inline src/arch/intsimdmatrixavx2.cpp:88]: Can `#if 1` and the `#else` part be removed?

Comment (robinwatts) [reply src/arch/intsimdmatrixavx2.cpp]: Fixed.

**Comment by robinwatts** [comment]
"make check" fails for me both on master and on my branch, so I can't check this properly. BUT... I've fixed the test code where it was calling Init() to pass in the scales vector too, so hopefully it'll work for you now.

**Comment by stweil** [comment]
I now tested with `test/testing/8087_054.3G.tif`, too, and get about 2 % faster processing:

    user	0m11.008s (old)
    user	0m10.810s (new)

The unittest now compiles, but fails when running `intsimmatrixtest`:

    $ ./intsimdmatrix_test 
    Running main() from ../../../../unittest/../googletest/googletest/src/gtest_main.cc
    [==========] Running 3 tests from 1 test suite.
    [----------] Global test environment set-up.
    [----------] 3 tests from IntSimdMatrixTest
    [ RUN      ] IntSimdMatrixTest.C
    ../../../../unittest/intsimdmatrix_test.cc:89: Failure
    Expected equality of these values:
      total
        Which is: -53.751.912
      -423243.392011
        Which is: -423.243,41
    [  FAILED  ] IntSimdMatrixTest.C (473 ms)
    [ RUN      ] IntSimdMatrixTest.SSE
    ../../../../unittest/intsimdmatrix_test.cc:89: Failure
    Expected equality of these values:
      total
        Which is: -53.751.912
      -423243.392011
        Which is: -423.243,41
    [  FAILED  ] IntSimdMatrixTest.SSE (437 ms)
    [ RUN      ] IntSimdMatrixTest.AVX2
    double free or corruption (out)

**Comment by stweil** [comment]
> "make check" fails for me both on master

It requires `tessdata`, `tessdata_best` and `tessdata_fast` on the same level as `tesseract`. In addition several font files must be available:

    $ ls -l test/testing/*ttf
    lrwxrwxrwx 1 stweil stweil    52 Jul  6 07:52 test/testing/ae_Arab.ttf -> /usr/share/fonts/truetype/fonts-arabeyes/ae_Arab.ttf
    lrwxrwxrwx 1 stweil stweil    61 Jul  6 07:48 test/testing/Arial_Bold_Italic.ttf -> /usr/share/fonts/truetype/msttcorefonts/Arial_Bold_Italic.ttf
    lrwxrwxrwx 1 stweil stweil    58 Jul  6 07:48 test/testing/DejaVuSans-ExtraLight.ttf -> /usr/share/fonts/truetype/dejavu/DejaVuSans-ExtraLight.ttf
    -rw-r--r-- 1 stweil stweil 81484 Sep  9  2009 test/testing/Lohit-Hindi.ttf
    lrwxrwxrwx 1 stweil stweil    59 Jul  6 07:49 test/testing/Times_New_Roman.ttf -> /usr/share/fonts/truetype/msttcorefonts/Times_New_Roman.ttf
    lrwxrwxrwx 1 stweil stweil    51 Jul  6 07:52 test/testing/UnBatang.ttf -> /usr/share/fonts/truetype/unfonts-core/UnBatang.ttf
    lrwxrwxrwx 1 stweil stweil    51 Jul  6 07:49 test/testing/Verdana.ttf -> /usr/share/fonts/truetype/msttcorefonts/Verdana.ttf

**Comment by robinwatts** [comment]
Make check doesn't even get to that stage for me. Compilation dies with:

In file included from apiexample_test.cc:32:0:
include_gunit.h:17:10: fatal error: fileio.h: No such file or directory
 #include "fileio.h"   // for tesseract::File
          ^~~~~~~~~~
compilation terminated.

Nonetheless, I've had another go at fixing intsimdmatrix_text.cpp. I had forgotten to scale the scales :)

**Comment by stweil** [comment]
> Compilation dies.

Did you run `make training` first? That is another requirement for `make check`.

**Comment by robinwatts** [comment]
With "make training" done first, it gets further, but then bitches about:

mkdir -p ../test/testing
ln -s /home/robin/sauce/ghostpdl/tesseract/test/testing/phototest.tif ../test/testing/phototest.tif
ln: failed to create symbolic link '../test/testing/phototest.tif': File exists
Makefile:2737: recipe for target '/home/robin/sauce/ghostpdl/tesseract/test/testing/phototest.tif' failed
make[2]: *** [/home/robin/sauce/ghostpdl/tesseract/test/testing/phototest.tif] Error 1

**Comment by stweil** [comment]
> With "make training" done first, it gets further, but then bitches ...

Do you have the `test/` directory with the `test` submodule? If not, run `git submodule update --init test`.

**Comment by robinwatts** [comment]
Yeah, I have that submodule (and all the rest, I believe!)

**Comment by robinwatts** [comment]
Apologies for the problems I'm having getting "make check" to work Stefan. I hope that the current branch should work OK for you though.

**Comment by zdenop** [comment]
@robinwatts @stweil, what is status of this PR?

**Comment by robinwatts** [comment]
I believe it's good to go. I just can't actually run "make check" to know that it passes. I've fixed the commit since Stefan last tried it and found that it failed.

**Comment by zdenop** [comment]
`make check` fails for intsimdmatrix_test with this PR (it works without it). Output (is quite long) from `intsimdmatrix_test` starts with :
```
Running main() from ../../../unittest/../googletest/googletest/src/gtest_main.cc
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from IntSimdMatrixTest
[ RUN      ] IntSimdMatrixTest.C
../../../unittest/intsimdmatrix_test.cc:84: Failure
Expected equality of these values:
  base_result[i]
    Which is: -343.81131
  test_result[i]
    Which is: -2.7071757
i=0
../../../unittest/intsimdmatrix_test.cc:84: Failure
Expected equality of these values:
  base_result[i]
    Which is: -2,619.0117
  test_result[i]
    Which is: -20.622139
i=0
../../../unittest/intsimdmatrix_test.cc:84: Failure
```
Unit tests run as :

```
git clone https://github.com/tesseract-ocr/tesseract.git
cd tesseract
git submodule update --init
git fetch origin pull/2994/head:pr2994
git checkout pr2994
./autogen.sh
mkdir -p bin/unittest && cd bin/unittest
../../configure --disable-shared 'CXXFLAGS=-g -O2 -Wall -Wextra -Wpedantic -fsanitize=address,undefined -fstack-protector-strong -ftrapv'
make training && make check
```

**Review by stweil — CHANGES_REQUESTED** [review]
`intsimdmatrix_test` fails for me, too (same errors as reported by @zdenop). Maybe the test code needs an update.

**Comment by robinwatts** [comment]
First off, thanks for your continued patience.

By following zdenop's script given above I can build and run the tesseract unittests. Or at least I can get further. With a checkout of master, I get 19 failures. With my changes brought up to date with master, and fixed, the intsimdmatrix unit test passes, but the same 19 other tests still fail. I think that's the best I can hope for.

The CI builds here also appear to be failing, but for nothing (that I can see) due to me.

**Comment by zdenop** [comment]
@egorpugin : CI build fails because of sw on Ubuntu and Mac

**Comment by egorpugin** [comment]
I'll reopen to check builds.

**Comment by zdenop** [comment]
@robinwatts : CI is fixed. Some test will fails because of [missing external data](https://github.com/tesseract-ocr/tesseract/tree/master/unittest#requirements). IMO It will be enough if you fix `intsimdmatrix_test`.

**Comment by robinwatts** [comment]
> IMO It will be enough if you fix `intsimdmatrix_test`.

It's fixed. intsimdmatrix_test works fine. All it needs is a review. :)

**Comment by zdenop** [comment]
thanks:

```
./imagedata_test
Running main() from ../../../unittest/../googletest/googletest/src/gtest_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from ImagedataTest
[ RUN      ] ImagedataTest.CachesProperly
Loaded 2/12 lines (1-2) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (3-4) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (9-10) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (5-6) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (7-8) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (12-12) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (11-12) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (10-11) of document ./tmp/documentdata0.lstmf
Loaded 4/12 lines (1-4) of document ./tmp/documentdata0.lstmf
Loaded 4/12 lines (9-12) of document ./tmp/documentdata0.lstmf
Loaded 4/12 lines (5-8) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (12-12) of document ./tmp/documentdata0.lstmf
Loaded 2/12 lines (11-12) of document ./tmp/documentdata0.lstmf
Loaded 3/12 lines (10-12) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (1-1) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (2-2) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (3-3) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (4-4) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (9-9) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (5-5) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (6-6) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (7-7) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (8-8) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (12-12) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (11-11) of document ./tmp/documentdata0.lstmf
Loaded 1/12 lines (10-10) of document ./tmp/documentdata0.lstmf
Loaded 12/12 lines (1-12) of document ./tmp/documentdata0.lstmf
[ERROR] Allowance = 2000000[ERROR] Allowance = 4000000[ERROR] Allowance = 1000000[ERROR] Allowance = 100000000[       OK ] ImagedataTest.CachesProperly (1,228 ms)
[ RUN      ] ImagedataTest.CachesMultiDocs
Loaded 3/6 lines (1-3) of document ./tmp/documentdata0.lstmf
Loaded 3/5 lines (1-3) of document ./tmp/documentdata1.lstmf
Loaded 3/7 lines (1-3) of document ./tmp/documentdata2.lstmf
Loaded 6/6 lines (1-6) of document ./tmp/documentdata0.lstmf
Loaded 5/5 lines (1-5) of document ./tmp/documentdata1.lstmf
Unloaded document ./tmp/documentdata0.lstmf, saving 6291456 memory
Loaded 2/5 lines (4-5) of document ./tmp/documentdata1.lstmf
Loaded 3/6 lines (4-6) of document ./tmp/documentdata0.lstmf
Loaded 7/7 lines (1-7) of document ./tmp/documentdata2.lstmf
Loaded 3/7 lines (4-6) of document ./tmp/documentdata2.lstmf
Unloaded document ./tmp/documentdata1.lstmf, saving 5242880 memory
Loaded 3/5 lines (1-3) of document ./tmp/documentdata1.lstmf
Loaded 6/6 lines (1-6) of document ./tmp/documentdata0.lstmf
Loaded 3/6 lines (1-3) of document ./tmp/documentdata0.lstmf
Unloaded document ./tmp/documentdata2.lstmf, saving 7340032 memory
Loaded 1/7 lines (7-7) of document ./tmp/documentdata2.lstmf
Loaded 5/5 lines (1-5) of document ./tmp/documentdata1.lstmf
Loaded 3/7 lines (1-3) of document ./tmp/documentdata2.lstmf
[INFO]  Page 0[INFO]  Page 1[INFO]  Page 2[INFO]  Page 3[INFO]  Page 4[INFO]  Page 5[INFO]  Page 6[INFO]  Page 7[INFO]  Page 8[INFO]  Page 9[INFO]  Page 10[INFO]  Page 11[INFO]  Page 12[INFO]  Page 13[INFO]  Page 14[INFO]  Page 15[INFO]  Page 16[INFO]  Page 17[INFO]  Page 18[INFO]  Page 19[INFO]  Page 20[INFO]  Page 21[       OK ] ImagedataTest.CachesMultiDocs (1,077 ms)
[----------] 2 tests from ImagedataTest (2,306 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (2,307 ms total)
[  PASSED  ] 2 tests.

```

**Comment by robinwatts** [comment]
Thanks!

**Comment by amitdo** [comment]
Thank for this improvement, but why did you recommitted all these individual patches that duplicates some of the latest commits instead of doing a merge on your side? IMO, this is very ugly and confusing.

**Comment by robinwatts** [comment]
> Thank for this improvement, but why did you recommitted all these individual patches that duplicates some of the latest commits instead of doing a merge on your side? IMO, this is very ugly and confusing.

The simple answer is that I didn't knowingly do this. I can only blame finger trouble while rebasing. Sorry.
If a project maintainer wants to rewrite history by force pushing, I won't be offended.

**Comment by amitdo** [comment]
>The simple answer is that I didn't knowingly do this

I understand and forgive :-)

**Comment by stweil** [comment]
@amitdo, @zdenop, should I do a forced push to rewrite the history?

Do we have some more performance data which compares old and new code? I don't see improved speed on Intel Core i5. @robinwatts, did you only test with Raspberry Pi?

**Comment by robinwatts** [comment]
I tested on both pi and an i7.

From above: "Running tesseract on 8087_054.3G.tif, runtime drops from 2.681s (real), 8.582 (user) to 2.582 (real), 8.174 (user) for identical results (times from an x86 box)."

That x86 box was an Intel Core i7 5930K.

**Comment by amitdo** [comment]
This PR was already merged, so force pushing to the PR's branch will do nothing I believe.

Force pushing to master is not a good idea.

**Comment by amitdo** [comment]
This PR was already merged, so force pushing to the PR's branch will do nothing I believe.

Force pushing to master is not a good idea.

**Comment by amitdo** [comment]
Testing on one image is not a good speed test IMO, due to the data loading overhead.

You can use this command:

`tesseract listoffiles out --oem 1 `

listoffiles:

```
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
path/to/file
```

For speed testing purpose, you can insert just one filename multiple times.

**Comment by stweil** [comment]
A test on 329 images with `ocrd-tesserocr-recognize` which only does text recognition on already detected lines (no overhead for task switches, model loading, layout recognition) shows an improvement of about 2.4 % (total time reduced from 21:28 to 20:58). The test was run on Intel Core i5.

File: unittest/intsimdmatrix_test.cc
Comment (stweil) [inline unittest/intsimdmatrix_test.cc:79]: Resizing `test_result` from `ro` to the smaller `num_out` causes a container overflow in `IntSimdMatrixTest.AVX2`. A similar container overflow occurs in `LSTMTrainerTest.BidiTest`.

Container overflows can be detected by building with `-stdlib=libc++`.

File: unittest/intsimdmatrix_test.cc
Comment (stweil) [inline unittest/intsimdmatrix_test.cc:76]: `base_result` only requires `num_out` elements and can be allocated with that size right from the beginning.

**Pull request merged**
